source/val/validation_state.cpp

```diff
@@ -254,9 +258,19 @@ bool ValidationState_t::IsOpcodeInCurrentLayoutSection(SpvOp op) {
 }
 
 DiagnosticStream ValidationState_t::diag(spv_result_t error_code) const {
+  return diag(error_code, instruction_counter_);
+}
+
+DiagnosticStream ValidationState_t::diag(spv_result_t error_code,
+                                         int instruction_counter) const {
+  std::string disassembly;
+  if (instruction_counter >= 0 && static_cast<size_t>(instruction_counter) <=
+                                      ordered_instructions_.size()) {
+    disassembly = Disassemble(ordered_instructions_[instruction_counter - 1]);
```

Comment:
This produces the right output in almost all situations.
Corner cases where it does not are where type info from prior instructions are required to correctly parse:  e.g. OpConstnat %int ....  and also OpSwitch 

That can wait for a future PR.  I'll file an issue for this.

Reply:
I filed #1627 but the results are better than I expected. :-)